CRM/Core/PrevNextCache/Sql.php

```diff
@@ -97,37 +116,39 @@ public function markSelection($cacheKey, $action, $ids = NULL) {
       if (is_array($ids)) {
         $cIdFilter = "(" . implode(',', $ids) . ")";
         $whereClause = "
-WHERE cachekey = %1
+WHERE cachekey = %1 AND expiry_date >= FROM_UNIXTIME(%3)
 AND (entity_id1 IN {$cIdFilter} OR entity_id2 IN {$cIdFilter})
 ";
       }
       else {
         $whereClause = "
-WHERE cachekey = %1
+WHERE cachekey = %1 AND expiry_date >= FROM_UNIXTIME(%3)
 AND (entity_id1 = %2 OR entity_id2 = %2)
 ";
         $params[2] = ["{$ids}", 'Integer'];
       }
       if ($action == 'select') {
         $whereClause .= "AND is_selected = 0";
-        $sql = "UPDATE civicrm_prevnext_cache SET is_selected = 1 {$whereClause}";
+        $sql = "UPDATE civicrm_prevnext_cache SET is_selected = 1, expiry_date = FROM_UNIXTIME(%3) {$whereClause}";
         $params[1] = [$cacheKey, 'String'];
       }
       elseif ($action == 'unselect') {
         $whereClause .= "AND is_selected = 1";
-        $sql = "UPDATE civicrm_prevnext_cache SET is_selected = 0 {$whereClause}";
+        $sql = "UPDATE civicrm_prevnext_cache SET is_selected = 0, expiry_date = FROM_UNIXTIME(%3) {$whereClause}";
```

Comment:
We shouldn't update the expiry date for one row within an 'expiry set'

Reply:
This seems to be the case for the select key in redis as per https://github.com/civicrm/civicrm-core/blob/master/CRM/Core/PrevNextCache/Redis.php#L111 @totten

Reply:
Sort of :)

* Consider a use-case for this bit of code: "*A user is browsing the result-set, checks some items, and then decides to "Clear all" selections. He then continues to select some other items.*" This operation is the "clear all selections".
* In the Redis.php example, it does reset the TTL, but that wasn't the main intent/goal. It was just side-effect that was hard to avoid.
    * The intent is to *clear out the cache-object which contains the full set of selected IDs* (`redis->delete($selKey)`). IIRC, this basically meant deleting it and re-creating it. To re-create the cache-object, one has to call `setTimeout()`.
    * It was a bit lazy/dirty on my part to use `self:TTL` for the re-created object b/c it effectively extends the timeout (beyond its original value). I rationalized that it's OK: the re-created `$selKey` will expire *after* the existing `$allKey`. We could save some memory by cleaning `$selKey` sooner, but we don't lose any functionality or data-integrity by keeping `$selKey` a bit longer. Why not? `$allKey` is the main driver of the screen. When `$allKey` expires, everything else is irrelevant.
* In the Sql.php example above, it extends the life of all data in the prevnext cache.
    * IMHO, it's OK either to keep the TTL (like it did before) or to renew it (like it does above). Renewing the TTL based on user-interaction basically gives the cache some LRU-ish characteristics. From a policy perspective, I think basic TTL or LRU or some arbitrary in-between thing will do the job. We tend to use TTL as a matter of KISS.